rustls/src/client/builder.rs

```diff
@@ -163,6 +162,35 @@ impl ConfigBuilder<ClientConfig, WantsClientCert> {
         Ok(self.with_client_cert_resolver(Arc::new(resolver)))
     }
 
+    /// Sets a raw public key and matching private key for use
+    /// in client authentication.
+    ///
+    /// `cert_chain` is a vector of a DER-encoded RPK certificate.
+    /// `key_der` is a DER-encoded private key as PKCS#1, PKCS#8, or SEC1. The
+    /// `aws-lc-rs` and `ring` [`CryptoProvider`]s support all three encodings,
+    /// but other `CryptoProviders` may not.
+    ///
+    /// This function fails if `key_der` is invalid, or if the public_key cannot be parsed to a RPK cert
+    pub fn with_client_auth_rpk(
+        self,
+        public_key_der: SubjectPublicKeyInfoDer<'static>,
```

Comment:
I would suggest removing this argument, and using `private_key.public_key()` below. That should remove a category of accidental errors in this API (mismatched private/public key).

Reply:
This would mean this API is only usable for `SigningKey` impls that can produce the corresponding public key. This covers all of them in the base crate, but it's worth mentioning this detail in the doc for this function if we go this route.

Reply:
I do like that approach, seems to be more robust. In that case we could remove `public_key_der` from `client::builder::with_client_auth_rpk()`, `server::builder::with_rpk()`, `AlwaysResolvesClientRawPublicKeys::new()` and `AlwaysResolvesServerRawPublicKeys::new()`.
Currently the `with_client_auth_rpk()/with_rpk()` functions verify whether or not the `public_key_der` can be read into a `RawPublicKeyEntity`. Now it's probably better to verify that `SigningKey::public_key()` does not return `None` right? 
Removing `public_key_der` from these functions might also make `get_client_spki()/client_spki()` in `tests::common::mod` redundant. If we would remove these I doubt there is much use for https://github.com/rustls/rustls/pull/2078.

Reply:
Thanks for the good suggestion, looks better now! As of now I feel that the changes made in  https://github.com/rustls/rustls/pull/2078 might not be necessary. But maybe you foresee testing scenarios where it is useful to have RPK pemfiles.